### PR TITLE
fix: Idle handler refactor

### PR DESCRIPTION
**Pull request by jy4096**
Issue: currently, when watermark is disabled, we still send control messages. The expected behavior is to skip control messages and watermark.
- Introduce a new Idle Manager Interface
- Change the idle manager interface method name and update the related logic
- Add idleManager as the input argument for all forwarders
- Update tests

Also: Any naming suggestion for IdleManager and IdleManagerInterface....??

### CODE REVIEW

File: pkg/watermark/generic/jetstream/generic.go [hunk 1]
```diff
@@ -115,9 +115,6 @@ func BuildSourcePublisherStores(ctx context.Context, vertexInstance *v1alpha1.Ve
 	if !vertexInstance.Vertex.IsASource() {
 		return nil, fmt.Errorf("not a source vertex")
 	}
-	if vertexInstance.Vertex.Spec.Watermark.Disabled {
```
Comment (whynowy) [inline pkg/watermark/generic/jetstream/generic.go:118]: why?

File: pkg/watermark/wmb/noop_idle_manager.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+import (
+	"github.com/numaproj/numaflow/pkg/isb"
+)
+
+type NoopIdleManager struct {
+}
+
+// NewNoopIdleManager returns an no op IdleManager object
+func NewNoopIdleManager() *NoopIdleManager {
```
Comment (whynowy) [inline pkg/watermark/wmb/noop_idle_manager.go:11]: `NoOp`

File: pkg/watermark/wmb/interface.go [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package wmb
+
+import "github.com/numaproj/numaflow/pkg/isb"
+
+// IdleManagerInterface decides when to send a control message and also keeps track of idle watermark's offset.
+type IdleManagerInterface interface {
+	// Validate validates whether to send a control message for the given partition or not
+	Validate(toBufferPartitionName string) bool
```
Comment (whynowy) [inline pkg/watermark/wmb/interface.go:8]: Naming..

`Validate` is a verb, using something like `IsNeeded`, `IsValid`, figure out a better name.

File: pkg/watermark/generic/jetstream/generic.go (hunk 1 above)
Comment (jy4096) [reply pkg/watermark/generic/jetstream/generic.go:118]: The Watermark.Idsabled condition check is now at the parent level
https://github.com/numaproj/numaflow/pull/1021/files#diff-685a4c165a187fb1d7c5f29325d1ac019894e70a42d915f69e394383c417b768R116

File: pkg/watermark/wmb/interface.go (hunk 3 above)
Comment (jy4096) [reply pkg/watermark/wmb/interface.go]: Updated to `NeedToSendCtrlMsg`... any other suggestions.... ?...

**Comment by vigith** [comment]
> Also: Any naming suggestion for IdleManager and IdleManagerInterface....??
`IdleManagement` for type `IdleManager` for interface?

File: pkg/sinks/forward/forward.go [hunk 4]
```diff
@@ -51,7 +51,7 @@ type DataForward struct {
 	vertexName          string
 	pipelineName        string
 	// idleManager manages the idle watermark status.
-	idleManager *wmb.IdleManager
+	idleManager wmb.IdleManagerInterface
```
Comment (vigith) [inline pkg/sinks/forward/forward.go:54]: type name is used for interface name, confusing

File: pkg/sinks/sink.go [hunk 5]
```diff
@@ -129,7 +130,10 @@ func (u *SinkProcessor) Start(ctx context.Context) error {
 
 			// create watermark publisher using watermark stores
 			publishWatermark = jetstream.BuildPublishersFromStores(ctx, u.VertexInstance, sinkWmStores)
+			// sink vertex has only one toBuffer, so the length is 1
+			idleManager = wmb.NewIdleManager(1)
```
Comment (vigith) [inline pkg/sinks/sink.go:134]: does sink need a publisher?

File: pkg/watermark/wmb/interface.go [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+import "github.com/numaproj/numaflow/pkg/isb"
+
+// IdleManagerInterface decides when to send a control message and also keeps track of idle watermark's offset.
+type IdleManagerInterface interface {
+	// NeedToSendCtrlMsg validates whether to send a control message for the given partition or not
+	NeedToSendCtrlMsg(toBufferPartitionName string) bool
+	Get(toBufferPartitionName string) isb.Offset
+	Update(toBufferPartitionName string, newOffset isb.Offset)
+	Reset(toBufferPartitionName string)
```
Comment (vigith) [inline pkg/watermark/wmb/interface.go:11]: please document these methods

**Review by whynowy — APPROVED** [review]

File: pkg/sinks/sink.go (hunk 5 above)
Comment (jy4096) [reply pkg/sinks/sink.go:135]: Yes because UI needs to fetch the sink watermark.

File: pkg/watermark/wmb/idle_manager.go [hunk 7]
```diff
@@ -36,11 +36,13 @@ func NewIdleManager(length int) *IdleManager {
 	}
 }
 
-// Exists returns true if the given toBuffer partition name exists in the IdleManager map.
-func (im *IdleManager) Exists(toBufferPartitionName string) bool {
+// NeedToSendCtrlMsg returns true if the given partition hasn't got any control message and needs to create a new control message
+func (im *IdleManager) NeedToSendCtrlMsg(toBufferPartitionName string) bool {
 	im.lock.RLock()
```
Comment (yhl25) [inline pkg/watermark/wmb/idle_manager.go:41]: make `IdleManager` private? also return interface when `NewIdleManager` is called?

File: pkg/watermark/wmb/interface.go [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package wmb
+
+import "github.com/numaproj/numaflow/pkg/isb"
+
+// IdleManagement decides when to send a control message and also keeps track of idle watermark's offset.
+type IdleManagement interface {
+	// NeedToSendCtrlMsg validates whether to send a control message for the given partition or not.
```
Comment (yhl25) [inline pkg/watermark/wmb/interface.go:7]:
```suggestion

// IdleManagement decides when to send a control message and also keeps track of idle watermark's offset.
type IdleManager interface {
	// NeedToSendCtrlMsg validates whether to send a control message for the given partition or not.
```
as per @vigith comment, it should be IdleManager.

**Review by yhl25 — CHANGES_REQUESTED** [review]

**Review by yhl25 — APPROVED** [review]

**Pull request merged**